File: Set1_1000/22_Generate_Parentheses.java
class Solution {
    List<String> result = new ArrayList();
    public List<String> generateParenthesis(int n) {
        backTrack("", 0, 0, n);
        return result;
    }
    public void backTrack(String s, int left, int right, int n){
        if (s.length() == 2 * n){
            result.add(s);
            return;
        }
        if (left < n)
            backTrack(s + '(', left+1, right, n);
        if (right < left)
            backTrack(s + ')', left, right+1, n);
    }
}

File: Set1_1000/22_Generate_Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        answer = []
        def backtrackSolution(string = '', left = 0, right = 0):
            if len(string) == 2 * n:
                answer.append(string)
                return 
            if left < n:
                backtrackSolution(string + '(', left + 1, right)
            if right < left:
                backtrackSolution(string + ')', left, right + 1)
                
        backtrackSolution()
        return answer
